### PR TITLE
Transparent batching of ZK operations

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

It is more efficient to do operations to ZK in batch, using the `ZooKeeper.multi()` operation instead of individual calls. 
This reduces the RPC traffic between ZK client and servers and it also reduces the number of write transactions, since each `multi()` correspond to a single ZK transaction, containing multiple write operations.

### Modifications

 * Added a base class `AbstractBatchedMetadataStore` which contains the batching logic
 * Added configurable batching parameters
 * Using 2 different transaction sets, one for read operations and the other just for write operations
 * Added code of (very) basic benchmark

### Preliminary results

This is done with the included basic benchmark, making request to the embedded ZK running on a laptop:

image

image

### CODE REVIEW

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java [hunk 1]
```diff
@@ -128,105 +143,227 @@ protected void receivedSessionEvent(SessionEvent event) {
...
-                    if (code == Code.OK) {
-                        future.complete(Optional.of(new GetResult(data, getStat(path1, stat))));
-                    } else if (code == Code.NONODE) {
-                        future.complete(Optional.empty());
+                    if (code == Code.CONNECTIONLOSS) {
+                        log.info("CONNECTION LOSS");
+                        // There is the chance that we caused a connection reset by sending or requesting a batch
+                        // that passed the max ZK limit. Retry with the individual operations
+                        executor.schedule(() -> {
+                            log.info("FUTURE IS TRIGGERED");
```
Comment (eolivelli) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java:156]: Nit: remove this log

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java [hunk 2]
```diff
@@ -128,105 +143,227 @@ protected void receivedSessionEvent(SessionEvent event) {
...
+                            break;
+                        case DELETE:
+                            handleDeleteResult(op.asDelete(), opr);
+                            break;
+                        case GET:
+                            handleGetResult(op.asGet(), opr);
+                            break;
+                        case GET_CHILDREN:
+                            handleGetChildrenResult(op.asGetChildren(), opr);
+                            break;
```
Comment (eolivelli) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java:183]: Add a default case with something that tells us about an illegal state

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                                               EnumSet<CreateOption> options) {
+        OpPut op = new OpPut(path, data, optExpectedVersion, options);
+        enqueue(writeOps, op);
+        return op.getFuture();
+    }
+
+    private void enqueue(MessagePassingQueue queue, MetadataOp op) {
+        if (enabled) {
+            if (!queue.offer(op)) {
+                op.getFuture().completeExceptionally(new IllegalStateException("metadata queue is full"));
```
Comment (eolivelli) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java:141]: Os it better to fail with an error of the Metadata Store Exceptions family?
This way caller code won't see surprises with RuntimeException

File: pulsar-metadata/src/test/java/org/apache/pulsar/metadata/MetadataBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+import lombok.Cleanup;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.common.util.FutureUtil;
+import org.apache.pulsar.metadata.api.MetadataStore;
+import org.apache.pulsar.metadata.api.MetadataStoreConfig;
+import org.apache.pulsar.metadata.api.MetadataStoreFactory;
+import org.testng.annotations.Test;
+
+@Slf4j
+public class MetadataBenchmark extends MetadataStoreTest {
```
Comment (eolivelli) [inline pulsar-metadata/src/test/java/org/apache/pulsar/metadata/MetadataBenchmark.java:37]: How many resources does this test use? How does it take to complete? 
Does it make sense to keep this running on CI?

I believe that if the test is no heavyweight we can keep it, but I wanted to double check with you

**Review by eolivelli — COMMENTED** [review]
Great work!

**Comment by lhotari** [comment]
Awesome!

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java [hunk 5]
```diff
@@ -128,105 +143,229 @@ protected void receivedSessionEvent(SessionEvent event) {
...
+
+                        default:
+                            op.getFuture().completeExceptionally(new IllegalStateException(
+                                    "Operation type not supported in multi: " + op.getType()));
+                    }
+                }
             }, null);
         } catch (Throwable t) {
-            future.completeExceptionally(new MetadataStoreException(t));
+            ops.forEach(o -> o.getFuture().completeExceptionally(t));
```
Comment (lhotari) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java:190]: Just wondering if t should be wrapped once in a MetadataStoreException to be consistent.

**Review by lhotari — COMMENTED** [review]
Nice work! One minor comment about exception handling.

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java (hunk 5 above)
Comment (merlimat) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/ZKMetadataStore.java:190]: 👍 Yes, it makes sense to wrap it

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    private void enqueue(MessagePassingQueue queue, MetadataOp op) {
+        if (enabled) {
+            if (!queue.offer(op)) {
+                // Execute individually if we're failing to enqueue
+                batchOperation(Collections.singletonList(op));
+                return;
+            }
+            if (queue.size() > maxOperations && flushInProgress.compareAndSet(false, true)) {
```
Comment (codelipenghui) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java:153]: We will not trigger a flush if reach the size limitation? Looks like the current implementation is split into multiple batches when flushing, if we have enough data should be flush not the operations, we will wait for `maxDelayMillis`.

Comment (codelipenghui) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java:153]: The change LGTM and just left a minor comment, I think it should not be a big problem here, but will simplify implementation.

**Review by codelipenghui — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]
LGTM

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java (hunk 6 above)
Comment (merlimat) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java:153]: You're right, there are several optimizations that are still possible in the batching logic :) I'd say to continue in subsequent PRs.

**Comment by Anonymitaet** [comment]
For the doc side, I've discussed with @codelipenghui, the following changes will be made after 2.9.0 is released:

1. Add parameters to the [Pulsar configuration page](https://pulsar.apache.org/docs/en/next/reference-configuration/#broker).

2. Add descriptions and two images in this PR (result comparison) to [ZooKeeper](https://pulsar.apache.org/docs/en/next/administration-zk-bk/#zookeeper).

@merlimat any thoughts? Thanks

**Pull request merged**